### PR TITLE
Feature/mv basho smoothing1

### PR DESCRIPTION
**Pull request by matthewvon**
Establish a smoothing algorithm to replace RocksDB default stall/stop strategy.

### CODE REVIEW

File: lib/ApplicationFeatures/RocksDBOptionFeature.cpp [hunk 1]
```diff
@@ -84,8 +84,9 @@ RocksDBOptionFeature::RocksDBOptionFeature(
     testSize >>= 1;
   }
   // setting the number of background jobs to
+  //  ... for edge case of 6 to 8 cpus, be sure to reserve one of 6 for I/O (mev)
   _maxBackgroundJobs = static_cast<int32_t>(std::max((size_t)2,
-                                std::min(TRI_numberProcessors(), (size_t)8)));
+                                                     std::min((TRI_numberProcessors()*5)/6, (size_t)12)));
```
[the remark this answers is not in the document]
Comment (matthewvon) [reply lib/ApplicationFeatures/RocksDBOptionFeature.cpp]: I am reverting this line in favor of the priority swapping in Linux build.  Might come back here and perform a simple "-1" for non-Linux environments.  But I have not started non-Linux adjustments.

File: arangod/RocksDBEngine/RocksDBThrottle.h [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+#include <db/db_impl.h>
+#include <db/write_controller.h>
+
+namespace arangodb {
+
+////////////////////////////////////////////////////////////////////////////////
+/// If these values change, make sure to reflect the changes in
+/// RocksDBPrefixExtractor as well.
+////////////////////////////////////////////////////////////////////////////////
+class RocksDBEventListener : public rocksdb::EventListener {
```
[the remark this answers is not in the document]
Comment (matthewvon) [reply arangod/RocksDBEngine/RocksDBThrottle.h]: Done.

File: arangod/RocksDBEngine/RocksDBThrottle.h [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  //  do not want sudden changes in one or two intervals to swing
+  //  the throttle value wildly.  Goal is a nice, even throttle value.
+  static const unsigned THROTTLE_SCALING = 17;
+
+  // trigger point where level-0 file is considered "too many pending"
+  //  (from original Google leveldb db/dbformat.h)
+  static const int64_t kL0_SlowdownWritesTrigger = 8;
+
+  struct ThrottleData_t
+  {
```
[the remark this answers is not in the document]
Comment (matthewvon) [reply arangod/RocksDBEngine/RocksDBThrottle.h]: Done

**Comment by matthewvon** [comment]
Key items still remaining:
1. change to arangodb mutex
2. create compile conditions for new Linux specific code, maybe add one change for non-Linux machines in lib/ApplicationFeatures
3. continue clean up of variable naming

This are just syntactical in nature, not functional.  I believe I am code complete at this point.  Smoothing works.  Secondary goal of better 4 cpu management now added.

File: arangod/RocksDBEngine/RocksDBThrottle.h [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  void SetThrottle();
+
+  int64_t ComputeBacklog();
+
+  void RecalculateThrottle();
+
+  // I am unable to figure out static initialization of std::chrono::seconds,
+  //  using old school unsigned.
+  static const unsigned THROTTLE_SECONDS = 60;
+  static const unsigned THROTTLE_INTERVALS = 63;
```
[the remark this answers is not in the document]
Comment (matthewvon) [reply arangod/RocksDBEngine/RocksDBThrottle.h]: now addressed

File: arangod/RocksDBEngine/RocksDBThrottle.h [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    std::chrono::microseconds m_Micros;
+    uint64_t m_Keys;
+    uint64_t m_Bytes;
+    uint64_t m_Compactions;
+  };
+
+  rocksdb::DBImpl * internalRocksDB_;
+  std::once_flag init_flag_;
+  std::atomic<bool> thread_running_;
+  std::future<void> thread_future_;
```
[the remark this answers is not in the document]
Comment (matthewvon) [reply arangod/RocksDBEngine/RocksDBThrottle.h]: Done

**Comment by matthewvon** [comment]
I believe all requested changes are done.  Since there are minor differences in ArangoDB's condition variable usage and std::condition_variable, I have a new test running ... and it seems fine so far after 30 minutes.

**Comment by jsteemann** [comment]
@matthewvon can you do me a favor and resolve the merge conflicts with latest devel?
After that, do you think the PR is ready to be merged back into devel?

**Comment by matthewvon** [comment]
@jsteemann I merged "devel" into this branch.  The conflict was stupid.  It should not have been a conflict ... but it was.

**Comment by jsteemann** [comment]
@matthewvon the Windows build seems to fail with
```
"E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\ALL_BUILD.vcxproj" (default target) (1) ->
"E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\arangod\arangod.vcxproj" (default target) (12) ->
"E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\arangod\arangoserver.vcxproj" (default target) (13) ->
(ClCompile target) -> 
  E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\3rdParty\rocksdb\v5.6.X\port/win/port_win.h(53): error C2371: 'ssize_t': redefinition; different basic types (compiling source file E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\arangod\RocksDBEngine\RocksDBEngine.cpp) [E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\arangod\arangoserver.vcxproj]
  E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\3rdParty\rocksdb\v5.6.X\port/win/port_win.h(53): error C2371: 'ssize_t': redefinition; different basic types (compiling source file E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\arangod\RocksDBEngine\RocksDBThrottle.cpp) [E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\arangod\arangoserver.vcxproj]
  E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\arangod\RocksDBEngine\RocksDBThrottle.cpp(37): fatal error C1083: Cannot open include file: 'sys/syscall.h': No such file or directory [E:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\arangod\arangoserver.vcxproj]

    6 Warning(s)
    3 Error(s)
```
Can you have a look at the feature branch's test results in Jenkins? Thanks!

**Comment by matthewvon** [comment]
Not surprised.  Windows build was never attempted.  That was somehow "secondary".  

Yes, looking at it now.

Matthew

**Comment by matthewvon** [comment]
Builds appears to work on Windows.

**Comment by jsteemann** [comment]
Note sure here. It looks to me like the `ssize_t redefinition` issues are still there:
```
"F:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\ALL_BUILD.vcxproj" (default target) (1) ->
"F:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\arangod\arangod.vcxproj" (default target) (12) ->
"F:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\arangod\arangoserver.vcxproj" (default target) (13) ->
(ClCompile target) -> 
  F:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\3rdParty\rocksdb\v5.6.X\port/win/port_win.h(53): error C2371: 'ssize_t': redefinition; different basic types (compiling source file F:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\arangod\RocksDBEngine\RocksDBEngine.cpp) [F:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\arangod\arangoserver.vcxproj]
  F:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\3rdParty\rocksdb\v5.6.X\port/win/port_win.h(53): error C2371: 'ssize_t': redefinition; different basic types (compiling source file F:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\arangod\RocksDBEngine\RocksDBThrottle.cpp) [F:\jenkins\workspace\arangodb-ci_PR-3384-MSVY3SZCGEFHCJ3IFM536IZ6755FUE7RDSWTUA3254SXTOWNL7VQ\build\arangod\arangoserver.vcxproj]

    6 Warning(s)
    2 Error(s)
```
http://c1.triagens-gmbh.zz:8080/view/3.features/job/arangodb-ci/view/change-requests/job/PR-3384/14//artifact/windows-enterprise-maintainer/01-build-FAIL/build.log

**Comment by matthewvon** [comment]
Did some deeper digging within rocksdb.  Latest change has potential, but I do not have a Windows box for direct verification.  Must wait on Jenkins.

**Comment by fceller** [comment]
The following crash occurred on shutdown

```
(gdb) where
#0  std::default_delete<rocksdb::RateLimiter>::operator() (this=<optimized out>, __ptr=0x255e12) at /usr/include/c++/5/bits/unique_ptr.h:76
#1  std::unique_ptr<rocksdb::RateLimiter, std::default_delete<rocksdb::RateLimiter> >::~unique_ptr (this=0x7f1b0b168420, 
    __in_chrg=<optimized out>) at /usr/include/c++/5/bits/unique_ptr.h:236
#2  rocksdb::DBImpl::~DBImpl (this=0x7f1b0b167c00, __in_chrg=<optimized out>)
    at /tmp/90bdb97a05d382785eebbe0d93105246/3rdParty/rocksdb/v5.6.X/db/db_impl.cc:249
#3  0x0000000001cca651 in rocksdb::DBImpl::~DBImpl (this=0x7f1b0b167c00, __in_chrg=<optimized out>)
    at /tmp/90bdb97a05d382785eebbe0d93105246/3rdParty/rocksdb/v5.6.X/db/db_impl.cc:349
#4  0x0000000001e17a0f in rocksdb::StackableDB::~StackableDB (this=0x7f1af0bbf340, __in_chrg=<optimized out>)
    at /tmp/90bdb97a05d382785eebbe0d93105246/3rdParty/rocksdb/v5.6.X/include/rocksdb/utilities/stackable_db.h:25
#5  rocksdb::TransactionDB::~TransactionDB (this=0x7f1af0bbf340, __in_chrg=<optimized out>)
    at /tmp/90bdb97a05d382785eebbe0d93105246/3rdParty/rocksdb/v5.6.X/include/rocksdb/utilities/transaction_db.h:151
#6  rocksdb::TransactionDBImpl::~TransactionDBImpl (this=0x7f1af0bbf340, __in_chrg=<optimized out>)
    at /tmp/90bdb97a05d382785eebbe0d93105246/3rdParty/rocksdb/v5.6.X/utilities/transactions/transaction_db_impl.cc:67
```

**Comment by fceller** [comment]
Core is unrelated

**Pull request merged**